### DISCUSSION TITLE
`paramManager` throws during upgrade

### DISCUSSION THREAD

**Discussion by anilhelvaci** [Agoric/agoric-sdk#8303] [Q&A] [answered]
## Context
I'm trying to run a governance election in swingset tests. I'm able to start my governed contract in the bootstrap without any problems. But when I pose a question using my contract governor and then try to upgrade I get the below error:

```sh
Logging sent error stack (Error#1)
Error#1: A Zoe invitation is required, not Object [Alleged: Zoe Invitation payment] {}
  at onRejected (.../zoe/src/zoeService/invitationQueries.js:10:18)

Error#1 ERROR_NOTE: Due to (Error#2)
Error#1 ERROR_NOTE: Sent as error:liveSlots:v6#70001
Nested error under Error#1
  Error#2: Object [Alleged: Zoe Invitation payment] {} was not a live payment for brand Object [Alleged: Zoe Invitation brand] {} . It could be a used-up payment, a payment for another brand, or it might not be a payment at all.
    at assertLivePayment (.../ertp/src/paymentLedger.js:234:6)
    at Object.getAmountOf (.../ertp/src/paymentLedger.js:351:1)
    at Alleged: Zoe Invitation issuer.method (.../exo/src/exo-tools.js:183:8)
    at eval (.../exo/src/exo-tools.js:91:8)
  
Logging sent error stack (Error#3)
Error#3: Object [Alleged: Zoe Invitation payment] {} was not a live payment for brand Object [Alleged: Zoe Invitation brand] {} . It could be a used-up payment, a payment for another brand, or it might not be a payment at all.
  at assertLivePayment (.../ertp/src/paymentLedger.js:234:6)
  at Object.getAmountOf (.../ertp/src/paymentLedger.js:351:1)
  at Alleged: Zoe Invitation issuer.method (.../exo/src/exo-tools.js:183:8)
  at eval (.../exo/src/exo-tools.js:91:8)

Error#3 ERROR_NOTE: Sent as error:liveSlots:v6#70002
🤞 UnhandledPromiseRejection: (RemoteError(error:liveSlots:v6#70001)#4)
RemoteError(error:liveSlots:v6#70001)#4: A Zoe invitation is required, not "[Alleged: Zoe Invitation payment]"
  at decodeErrorCommon (.../marshal/src/marshal.js:281:20)
  at decodeErrorFromSmallcaps (.../marshal/src/marshal.js:329:8)
  at decodeFromSmallcaps (.../marshal/src/encodeToSmallcaps.js:435:14)
  at Object.fromCapData (.../marshal/src/marshal.js:356:15)
  at notifyOnePromise (.../swingset-liveslots/src/liveslots.js:1193:13)
  at notify (.../swingset-liveslots/src/liveslots.js:1217:15)
  at dispatchToUserspace (.../swingset-liveslots/src/liveslots.js:1486:1)
  at eval (.../swingset-vat/src/kernel/dummyMeterControl.js:39:31)
  at runWithoutMetering (.../swingset-vat/src/kernel/dummyMeterControl.js:21:8)
  at wrapped (.../swingset-vat/src/kernel/dummyMeterControl.js:39:8)

🤞 UnhandledPromiseRejection: (RemoteError(error:liveSlots:v6#70001)#4)
```

I use [committee.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/governance/src/committee.js) as my Electorate so I know the contract governor uses the `initialPoserInvitation` to call [addQuestion](https://github.com/Agoric/agoric-sdk/blob/fb2e7f5dec80c6bf17a16483fb3b956fced41029/packages/governance/src/committee.js#L134). So I think it should be expected that `initialPoserInvitation` is going to be used up when a question is asked. But when I upgrade my contract [paramManager.js](https://github.com/Agoric/agoric-sdk/blob/fb2e7f5dec80c6bf17a16483fb3b956fced41029/packages/governance/src/contractGovernance/paramManager.js#L257) tries to assert the payment is live. **My test still passes but I'm not sure if is it okay for me to ignore this issue for now?**

I'm aware of this pr https://github.com/Agoric/agoric-sdk/pull/8157 but I can't tell if it covers this issue. If not, I think we should consider at least documenting this.

cc @Chris-Hibbert @dckc

**Comment by anilhelvaci** [comment]
## Update
Just noticed that  I'm not able to ask any new questions after the upgrade. This might be important. I think the situation is related to below peace of code in the [paramManager.js](https://github.com/Agoric/agoric-sdk/blob/fb2e7f5dec80c6bf17a16483fb3b956fced41029/packages/governance/src/contractGovernance/paramManager.js#L305-L312):

```js
const finishInvitationParam = E.when(
      prepareToSetInvitation(invitation),
      invitationAndAmount => {
        setInvitation(invitationAndAmount);
        // delay until currentAmount is defined because readers expect a valid value
        namesToParams.init(name, publicMethods);
      },
    );
```

Since `prepareToSetInvitation` fails, the onFullfilled callback is not invoked. Therefore `setInvitation` is not called and the [contractGovernorKit.js](https://github.com/Agoric/agoric-sdk/blob/fb2e7f5dec80c6bf17a16483fb3b956fced41029/packages/governance/src/contractGovernorKit.js#L62) cannot get the `poserFacet`. Should we open an issue for this?

**Comment by dckc** [comment]
Note #8157 and the issues it references.

**Comment by anilhelvaci** [comment]
> I'm aware of this pr https://github.com/Agoric/agoric-sdk/pull/8157 but I can't tell if it covers this issue. If not, I think we should consider at least documenting this.

Yeah I'm aware of this PR but wasn't sure this particular problem is being addressed or not. Thanks @dckc

**Reply by dckc** [reply]
oh. I guess I read too fast.

I'm not sure docs.agoric.com covers upgrade at all, so for now, #5200 seems to document the fact that governance isn't durable.

**Comment by Chris-Hibbert** [comment, marked as answer]
Thanks for pointing this out, @anilhelvaci. I'll ensure that this is addressed with the rest of #8157.
